project/src/main/kotlin/org/ionproject/core/accessControl/Cache.kt

```diff
@@ -0,0 +1,103 @@
+package org.ionproject.core.accessControl
+
+import org.ionproject.core.accessControl.pap.entities.PolicyEntity
+import org.ionproject.core.accessControl.pap.entities.TokenEntity
+import org.ionproject.core.accessControl.pap.sql.AuthRepoImpl
+import org.ionproject.core.common.LogMessages
+import org.ionproject.core.common.customExceptions.UnauthenticatedUserException
+import org.ionproject.core.common.interceptors.LoggerInterceptor
+import org.ionproject.core.common.transaction.DataSourceHolder
+import org.ionproject.core.common.transaction.TransactionManagerImpl
+import org.slf4j.LoggerFactory
+import org.springframework.cache.CacheManager
+import org.springframework.stereotype.Component
+
+/**
+ *  Cache used to avoid having to check database everytime
+ *  a request is made and needs authentication.
+ */
+@Component
+class AccessControlCache {
+
+    private val logger = LoggerFactory.getLogger(LoggerInterceptor::class.java)
+    private val pap: AuthRepoImpl = AuthRepoImpl(TransactionManagerImpl(DataSourceHolder))
+
+    companion object {
+        private val cacheManager: CacheManager = CaffeineConfiguration().cacheManager()
+        private val tokenCache = cacheManager.getCache(CaffeineConfiguration.tokenCache)
+        private val policiesCache = cacheManager.getCache(CaffeineConfiguration.policiesCache)
+    }
+
+    /**
+     * saves the requested token by the hash for future
+     * requests.
+     */
+    fun getToken(tokenHash: String, derived: Boolean) : TokenEntity {
+
+        val cacheValue = tokenCache?.get(tokenHash)
```

Comment:
This seems to have a check-then-act issue, where multiple requests can still do the DB query. I believe the caffeine cache has support for that, i.e., block as requests while a single one is doing the query.